subscription.go

```diff
@@ -0,0 +1,219 @@
+			DiscardOldest:    true,
+			Filter:           nil,
+			QueueSize:        10,
+			SamplingInterval: 0.0,
+		},
+	}
+}
+
+type PublishNotificationData struct {
+	SubscriptionID uint32
+	Error          error
+	Value          interface{}
+}
+
+// Cancel() deletes the Subscription from Server and makes the Client forget it so that publishing
+// loops cannot deliver notifications to it anymore
+func (s *Subscription) Cancel() error {
+	delete(s.c.subscriptions, s.SubscriptionID)
+
+	req := &ua.DeleteSubscriptionsRequest{
+		SubscriptionIDs: []uint32{s.SubscriptionID},
+	}
+	var res *ua.DeleteSubscriptionsResponse
+	err := s.c.Send(req, func(v interface{}) error {
+		return safeAssign(v, &res)
+	})
+	if err != nil {
+		return err
+	}
+	if res.ResponseHeader.ServiceResult != ua.StatusOK {
+		return res.ResponseHeader.ServiceResult
+	}
+
+	return nil
+}
+
+func (s *Subscription) Monitor(ts ua.TimestampsToReturn, items ...*ua.MonitoredItemCreateRequest) (*ua.CreateMonitoredItemsResponse, error) {
+	// Part 4, 5.12.2.2 CreateMonitoredItems Service Parameters
+	req := &ua.CreateMonitoredItemsRequest{
+		SubscriptionID:     s.SubscriptionID,
+		TimestampsToReturn: ts,
+		ItemsToCreate:      items,
+	}
+
+	var res *ua.CreateMonitoredItemsResponse
+	err := s.c.Send(req, func(v interface{}) error {
+		return safeAssign(v, &res)
+	})
+	return res, err
+}
+
+func (s *Subscription) Unmonitor(monitoredItemIDs ...uint32) (*ua.DeleteMonitoredItemsResponse, error) {
+	req := &ua.DeleteMonitoredItemsRequest{
+		MonitoredItemIDs: monitoredItemIDs,
+		SubscriptionID:   s.SubscriptionID,
+	}
+	var res *ua.DeleteMonitoredItemsResponse
+	err := s.c.Send(req, func(v interface{}) error {
+		return safeAssign(v, &res)
+	})
+	return res, err
+}
+
+func publishOnceForSubscription(s *Subscription, acks []*ua.SubscriptionAcknowledgement) (*ua.PublishResponse, error) {
+	if acks == nil {
+		acks = []*ua.SubscriptionAcknowledgement{}
+	}
+	req := &ua.PublishRequest{
+		SubscriptionAcknowledgements: acks,
+	}
+
+	var res *ua.PublishResponse
+	err := s.c.sendWithTimeout(req, s.getPublishTimeout(), func(v interface{}) error {
+		return safeAssign(v, &res)
+	})
+	return res, err
+}
+
+func (s *Subscription) getPublishTimeout() time.Duration {
+	keepaliveIntervalMs := uint64(s.RevisedPublishingInterval) *
+		uint64(s.RevisedMaxKeepAliveCount)
+	if keepaliveIntervalMs > math.MaxUint32 {
+		keepaliveIntervalMs = math.MaxUint32
+	}
+	requestedTimeout := time.Duration(keepaliveIntervalMs) * time.Millisecond
+	if requestedTimeout < s.c.cfg.RequestTimeout {
+		return s.c.cfg.RequestTimeout
+	}
+	return requestedTimeout
+}
+
+// Run() starts an infinite loop that sends PublishRequests and delivers received
+// notifications to registered Subscriptions.
+func (s *Subscription) Run(ctx context.Context) {
+	var acks []*ua.SubscriptionAcknowledgement
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			res, err := publishOnceForSubscription(s, acks)
+			if err != nil {
```

Comment:
I've tried to refactor this with different approaches and this is the best I could come up with.

```go
// send the next publish request
// note that res contains data even if an error was returned
res, err := s.publish(acks)

switch {
case err == ua.StatusBadSequenceNumberUnknown:
	// At least one ack has been submitted repeatedly.
	// Ignore the error. Acks will be cleared below
case err == ua.StatusBadTimeout:
	// ignore and continue the loop
case err == ua.StatusBadNoSubscription:
	// Subscription already deleted and the publishing loop is still running
	// but will be stopped shortly.
	// todo(fs): how?

case err != nil:
	s.c.notifySubscriptions(ctx, err)
}

if res != nil {
...
```

Reply:
There needs to be another exit condition on this loop as well.  On a critical error (I encountered a SessionIDInvalid), the loop never exits because it's only waiting for a context close.  I'm thinking `s.c.notifySubscriptions(ctx, err)` should be followed by a `return`

Reply:
Do you think we should assume that any error we didn't explicitly ignore is irrecoverable? I tend to agree, but I'm not very sure.